crates/ark/src/lsp/completions/sources/composite.rs

```diff
@@ -78,16 +101,81 @@ pub(crate) fn get_completions(
         )?;
     }
 
-    // Remove duplicates
-    let mut uniques = HashSet::new();
-    completions.retain(|x| uniques.insert(x.label.clone()));
+    // Simplify to plain old CompletionItems and sort them
+    let completions = finalize_completions(&completions);
+
+    Ok(Some(completions))
+}
+
+fn push_completions<S>(
+    source: S,
+    completion_context: &CompletionContext,
+    completion_map: &mut HashMap<CompletionKey, CompletionWithSource>,
+) -> anyhow::Result<()>
+where
+    S: CompletionSource,
+{
+    let source_name = source.name();
+
+    if let Some(source_completions) = collect_completions(source, completion_context)? {
+        for item in source_completions {
+            let key = CompletionKey::from(&item);
```

Comment:
This doesn't feel right because the `from()` method here is basically just cloning the underlying values so we end up with them in two places - one in `CompletionKey` and one in `CompletionItem`

Ah, maybe a better idea is to implement `Hash` for `CompletionWithSource` manually in such a way that it hashes the `label` and maybe the `insert_text` if we still think that is right, and just use a `HashSet` again.